### DISCUSSION TITLE
Cursor animations and other visual effects

### DISCUSSION THREAD

**Discussion by valpackett** [wezterm/wezterm#1136] [Ideas]
https://github.com/neovide/neovide has really cool cursor effects, would be nice to have them here.

There's also various silly fun stuff around like [power mode](https://marketplace.visualstudio.com/items?itemName=hoovercj.vscode-power-mode) :D

**Comment by wez** [comment]
While I'm not personally sold on the warping cursor effect in neovide (the motion feels excessive and distracting to me, but I know this is subjective!), I have been thinking about eg: cursor animation.

Since wezterm supports frame scheduling for animated gifs/pngs, it's not a huge stretch to replace the currently separate logic that manages what are effectively frames for cursor blinking and consolidate it with the more general image frame management.

In other words, it wouldn't be terribly difficult to support using animated gifs for the text cursor instead of the current static textures.
The main reason I didn't jump on this already is that I don't know of a supply of gifs that are really usable for the cursor to make a great feature demo!
https://github.com/liamg/darktile does something like this, but using a static duck image.

The basic warp/stretch effect in neovide is probably not too difficult to replicate if wezterm could remember the prior cursor position and then interpolate the new coordinates for the quad when it notices a change, but we don't currently have a thing to manage that kind of animation: only timing for switching to a new frame, so the GUI layer would need to grow something to help with that, and in turn, that might be sufficient to use for some of the particle effects on the cursor trail.

**Reply by qerplunk** [reply]
Something like darktile's static image as a cursor is exactly what I was looking for. 

However, if this were to be implemented, would there be a way to only use the image on, for example, normal mode in vim, and keep the default line for insert mode?

**Reply by wez** [reply]
I'm going to use https://github.com/wez/wezterm/issues/1432 to track this as an issue rather than a discussion, as those cross-link a bit better when referenced by number.  I've sketched out some thoughts on where to go for the main animation bits themselves, but not put down any thoughts about eg: warping effects.

**Comment by scy** [comment]
I totally understand that the cursor warping can feel “excessive and distracting”, but when I first saw it in Neovide, I thought “wow, that’s really _helpful_!”. Sure, Neovide’s particle effects are just for showing off, but a warping cursor _can_ be an accessibility feature, just like smooth scrolling. It enables you to better track where the cursor is moving to, e.g. when jumping between tmux windows, or after a `Ctrl-O` in Vim.

On the other hand, I don’t care at all for having a Darktile-style duck image as my cursor. ;)

**Reply by vvtommy** [reply]
Hahaha, I knew someone would bring this up.
What I mean is...that duck cursor thing.

**Comment by Shock9616** [comment]
Just indicating interest in this as well! Neovide-like animations would be really nice for tracking large cursor movements

**Comment by ram-xv** [comment]
Super interested in this!!

@wez, any updates on this?

**Comment by benomahony** [comment]
Looks like kitty have implemented this here: https://github.com/kovidgoyal/kitty/pull/7970/files

Could we bring this through from there or use this as a reference implementation?

**Reply by tylxm** [reply]
After tried the kitty's animated cursor, I can't use other terminals anymore, it's not a distraction at all and it helps me locate the cursor easily.

**Reply by weeebdev** [reply]
> After tried the kitty's animated cursor, I can't use other terminals anymore, it's not a distraction at all and it helps me locate the cursor easily.

same thing here. Since I've been using zellij, basically I lost nothing.

**Reply by Shock9616** [reply]
I also tried it and loved it, I just don't want to have to port my Wezterm config to kitty + tmux/zellij. I like the simplicity of the Wezterm multiplexer and it integrates really nicely with my Neovim config

**Reply by hwamil** [reply]
I actually tried kitty but since they don't support ligatures all the filenames in yazi got messed up so I'm sticking with wezterm.

The cursor is really nice and I want it (badly lol) but wezterm just gives me that fuzzy wuzzy feeling that other terminals do not.

**Reply by TRPB** [reply]
I'd love to see this in wezterm. Having tried it, the kitty implementation is nowhere near as nice as Neovide:

- It only affects > 1 character movement. If you move up/down or move left/right a single character nothing happens
- It is obviously a trail, the actual cursor moves into position instantly rather than animating from/to the start/end position
- It doesn't affect cursor shape: Try leaving/entering insert mode in neovide, the cursor shape animates!
- As you type in neovide the cursor animates the one character movement 
- There is zero antialising, the animated cursor is jagged as hell 

There's definitely scope for getting this as good as Neovide in wezterm, having used Neovide for a while the typing experience is just beautiful with the animated cursor as you type

To use an approach similar to the kitty PR but with the points above, you'd need to, on movement:

1. Read the existing cursor shape/position
2. Hide the terminal cursor
3. Show the trail at the original position
4. Read the real hidden cursor position
5. Animate the trail to the new position
6. When the animation finishes, show the terminal cursor and hide the trail

Alternatively, you could show only the trail 100% of the time and just leave it still when not animating. The original cursor would need to be hidden all the time

**Reply by ErrorNoInternet** [reply]
There are configuration options to tweak kitty's cursor trail: https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.cursor_trail

**Reply by TRPB** [reply]
yeah, none of them solve the issues I mentioned above though. Try using Neovide and compare

edit: Actually, these options get close enough:

```
cursor_shape block
cursor_trail 1
cursor_trail_start_threshold 0
shell_integration no-cursor
```

would be nice to get some antialiasing though
